core/src/schedule.rs

```diff
@@ -0,0 +1,180 @@
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::ops::RangeFrom;
+
+use bimap::BiBTreeMap;
+use granular_id::{GranularId, UpperBounded};
+use topological_sort::TopologicalSort;
+
+use crate::element::GranId;
+use crate::variables::{VarAccess, Variable};
+use crate::{Context, Element, OutputFormat};
+
+type ScheduleId = usize;
+
+// In short, each element is identified by its "GranId". We can use these Ids in our sorting, and
+// then get the next ID to be evaluated. But when a parent node is evaluated, all its children
+// is invalid and any returned IDs from its children (which is "new") may previously have been
+// assigned to the previous children (which doesn't exist anymore). This means that we need to be
+// able to "remove" IDs, but that isn't possible in TopologicalSort. For this reason, we assign
+// new unique IDs to each GranId, of the type ScheduleId. We have an BiBTreeMap to keep track of
+// this mapping.
+
+pub struct Schedule {
```

Comment:
Not sure about this once, but we may not want to expose the schedule as part of the public API and make these `pub(crate)` instead. Or maybe the schedule module is never exposed anyway.

Reply:
I think the schedule module is never exposed, but I will make it pub(crate)!